### PR TITLE
Fix chat channels order and colors at login #622

### PR DESCRIPTION
**Pull request by LenweSaralonde**

### CODE REVIEW

File: totalRP3/core/impl/communication_protocol_broadcast.lua [hunk 1]
```diff
@@ -271,24 +271,65 @@ local function hideBroadcastChannelFromChatFrame()
...
-			local _, channelName = GetChannelName(channelIndex);
-			if channelName == broadcastChannelName then
-				(SwapChatChannelByLocalID or C_ChatInfo.SwapChatChannelsByChannelIndex)(channelIndex, channelIndex + 1);
+		local broadcastChannelIndex = GetChannelName(config_BroadcastChannel());
+		if broadcastChannelIndex == nil then return end
+
+		-- Get the index of the last channel
+		local lastChannelIndex = 0;
+		for index = broadcastChannelIndex, MAX_WOW_CHAT_CHANNELS do
+			if (C_ChatInfo.GetChannelShortcut(index) or "") ~= "" then
```
Comment (Meorawr) [inline totalRP3/core/impl/communication_protocol_broadcast.lua:288]: Split this onto a separate line to make it a bit clearer.

```suggestion
			local shortcut = C_ChatInfo.GetChannelShortcut(index);
			if not shortcut or shortcut == "" then
```

File: totalRP3/core/impl/communication_protocol_broadcast.lua [hunk 2]
```diff
@@ -271,24 +271,65 @@ local function hideBroadcastChannelFromChatFrame()
...
+	-- The channel list is empty
+	if #{GetChannelList()} == 0 then
+		return false;
+	end
+
+	-- Find gaps in the channel list
+	local hasGaps = false;
+	local previousIndex = 0;
+	for index = 1, MAX_WOW_CHAT_CHANNELS do
+		if (C_ChatInfo.GetChannelShortcut(index) or "") ~= "" then
```
Comment (Meorawr) [inline totalRP3/core/impl/communication_protocol_broadcast.lua:320]: Same here.

```suggestion
		local shortcut = C_ChatInfo.GetChannelShortcut(index);
		if not shortcut or shortcut == "" then
```

File: totalRP3/core/impl/communication_protocol_broadcast.lua [hunk 3]
```diff
@@ -271,24 +271,65 @@ local function hideBroadcastChannelFromChatFrame()
...
+
+		hideBroadcastChannelFromChatFrame();
 	end
 end
 
+--- Return true if the channel list is ready for the broadcast channel to join.
+-- @return isReady (boolean)
+local function isChannelListReady()
+	-- The channel list is empty
+	if #{GetChannelList()} == 0 then
```
Comment (Meorawr) [inline totalRP3/core/impl/communication_protocol_broadcast.lua:312]: No need to make a table for this.

```suggestion
	if select("#", GetChannelList()) == 0 then
```

File: totalRP3/core/impl/communication_protocol_broadcast.lua [hunk 4]
```diff
@@ -320,20 +361,27 @@ Comm.broadcast.init = function()
 			-- We'll send out the event nice and early to say we're setting up.
 			TRP3_API.events.fireEvent(TRP3_API.events.BROADCAST_CHANNEL_CONNECTING);
 
+			-- Allow to join the broadcast channel even if the channel list is not ready
+			-- after 10 seconds.
+			local canJoinWhenNotReady = false;
+			C_Timer.After(10, function() canJoinWhenNotReady = true end);
```
Comment (Meorawr) [inline totalRP3/core/impl/communication_protocol_broadcast.lua:367]: `canJoinWhenNotReady` is a bit of an unclear name.

```suggestion
			-- Force joining the broadcast channel if we wait too long.
			local forceJoinChannel = false;
			C_Timer.After(10, function() forceJoinChannel = true; end);
```

File: totalRP3/core/impl/communication_protocol_broadcast.lua [hunk 5]
```diff
@@ -320,20 +361,27 @@ Comm.broadcast.init = function()
...
+
 			local firstTime = true;
 			ticker = C_Timer.NewTicker(1, function(_)
 				if firstTime then firstTime = false; return; end
 				if GetChannelName(string.lower(config_BroadcastChannel())) == 0 then
-					Log.log("Step 1: Try to connect to broadcast channel: " .. config_BroadcastChannel());
-					JoinChannelByName(string.lower(config_BroadcastChannel()));
+					if canJoinWhenNotReady or isChannelListReady() then
```
Comment (Meorawr) [inline totalRP3/core/impl/communication_protocol_broadcast.lua:373]: Fix up the renamed variable.

```suggestion
					if forceJoinChannel or isChannelListReady() then
```

File: totalRP3/core/impl/communication_protocol_broadcast.lua [hunk 6]
```diff
@@ -271,24 +271,65 @@ local function hideBroadcastChannelFromChatFrame()
...
+			if index ~= previousIndex + 1 then
+				hasGaps = true;
+				break;
+			end
+			previousIndex = index;
+		end
+	end
+
+	-- Consider the channel list is ready if there is no gap
+	return not(hasGaps);
```
Comment (Meorawr) [inline totalRP3/core/impl/communication_protocol_broadcast.lua:330]: Remove the useless parentheses.

```suggestion
	return not hasGaps;
```

**Review by Meorawr — CHANGES_REQUESTED** [review]

**Comment by LenweSaralonde** [comment]
Don't merge, code change requests have errors.

File: totalRP3/core/impl/communication_protocol_broadcast.lua (hunk 1 above)
Comment (LenweSaralonde) [reply totalRP3/core/impl/communication_protocol_broadcast.lua]: This change inverts the logic. The if checks for a non empty shortcut (not nil and not empty string).
Correct change is :
```lua
local shortcut = C_ChatInfo.GetChannelShortcut(index) or "";
if shortcut ~= "" then
```

File: totalRP3/core/impl/communication_protocol_broadcast.lua (hunk 2 above)
Comment (LenweSaralonde) [reply totalRP3/core/impl/communication_protocol_broadcast.lua]: Same here, logic is inverted

**Review by Solanya — APPROVED** [review]

**Review by Meorawr — APPROVED** [review]

**Pull request merged**